pkg/dep/git.go

```diff
@@ -0,0 +1,163 @@
+	_ starlark.HasAttrs = (*GitRepo)(nil)
+	// asserts *GitRepo implements loader.Dependency interface.
+	_ loader.Dependency = (*GitRepo)(nil)
+
+	// RequiredFields is the list of required fields to initialize a GitRepo target.
+	RequiredFields = []string{NameKey, RemoteKey, CommitKey}
+)
+
+// GitRepo represents Isopod module source as remote git repo.
+type GitRepo struct {
+	*AbstractDependency
+	name, remote, commit string
+}
+
+// NewGitRepoBuiltin creates a new git_repository built-in.
+func NewGitRepoBuiltin() *starlark.Builtin {
+	return starlark.NewBuiltin(
+		"git_repository",
+		func(t *starlark.Thread, b *starlark.Builtin, args starlark.Tuple, kwargs []starlark.Tuple) (starlark.Value, error) {
+			absDep, err := NewAbstractDependency("git_repository", RequiredFields, kwargs)
+			if err != nil {
+				return nil, err
+			}
+			name, remote, commit, err := nameRemoteCommit(absDep)
+			if err != nil {
+				return nil, fmt.Errorf("cannot read params: %v", err)
+			}
+			gitRepo := &GitRepo{absDep, name, remote, commit}
+			loader.Register(gitRepo)
+			return gitRepo, nil
+		},
+	)
+}
+
+// Name returns the name of this git repo target.
+func (g *GitRepo) Name() string {
+	return g.name
+}
+
+// LocalDir returns the path to the directory storing the source.
+func (g *GitRepo) LocalDir() string {
+	return filepath.Join(Workspace, g.name, g.commit)
+}
+
+// Fetch is part of the Dependency interface.
+// It downloads the source of this dependency.
+func (g *GitRepo) Fetch() error {
+	outputDir := g.LocalDir()
+	if _, err := Shellf(gitCloneShellScript(outputDir, g.remote, g.commit)); err != nil {
+		return fmt.Errorf("failed to clone git repo `%v': %v", g.name, err)
+	}
+	return nil
+}
+
+func nameRemoteCommit(absDep *AbstractDependency) (name, remote, commit string, err error) {
+	if name, err = stringFromValue(absDep.Attrs[NameKey]); err != nil {
+		return
+	}
+	if remote, err = stringFromValue(absDep.Attrs[RemoteKey]); err != nil {
+		return
+	}
+	if commit, err = stringFromValue(absDep.Attrs[CommitKey]); err != nil {
+		return
+	}
+	return
+}
+
+func stringFromValue(v starlark.Value) (string, error) {
+	if v == nil {
+		return "", errors.New("nil value")
+	}
+	s, ok := v.(starlark.String)
+	if !ok {
+		return "", fmt.Errorf("%v is not a starlark string (got a `%s')", v, v.Type())
+	}
+	return string(s), nil
+}
+
+// Shellf execute the given shell command and wait until it finishes. Then
+// return the combined stdout and stderr, and error if any.
+func Shellf(format string, a ...interface{}) (string, error) {
+	s := fmt.Sprintf(format, a...)
+	log.V(1).Infof("Executing shell command: %v\n", s)
+	bytes, err := exec.Command("sh", "-c", s).CombinedOutput()
+	log.V(1).Infof("Shell command `%s' finished:\n%s", s, string(bytes))
+	return string(bytes), err
+}
+
+// gitCloneBashScript is a bash script to fetch Git files of a repo at a given commit sha, to a given path.
+func gitCloneShellScript(outputDir, gitRemoteURL, gitCommitSHA string) string {
```

Comment:
This function would have been a good use case for go templates.

Reply:
Not a super big fan of having this as a shell script... It adds a non-implicit dependency for `git` and makes it a little less usable in docker containers. Would be great to use a go git client.
But I understand that it's complicated including authentication and all...

Reply:
I refactored this to use go templates.

Good point about the implicit dependency on `git`. Authentication is not impossible with `go-git` package but using `git` command allows Isopod to use existing auth creds, which I think is a huge UX improvement.